token/src/token.rs

```diff
@@ -145,22 +143,34 @@ mod tests {
         };
         // TBD: figure out better + more robust signing mechanism
         let gens = PedersenGens::default();
-        Prover::build_tx(program, header, &bp_gens, |t, verification_keys| {
-            let signtx_keys: Vec<Scalar> = verification_keys
-                .iter()
-                .filter_map(|vk| {
-                    for k in &keys {
-                        if (k * gens.B).compress() == *vk.as_compressed() {
-                            return Some(*k);
-                        }
+        let utx = Prover::build_tx(program, header, &bp_gens)?;
+
+        // find all the secret scalars for the pubkeys used in the VM
+        let privkeys: Vec<Scalar> = utx
+            .signing_instructions
+            .iter()
+            .filter_map(|(pubkey, _msg)| {
+                for k in keys.iter() {
+                    if (k * gens.B).compress() == *pubkey.as_compressed() {
+                        return Some(*k);
                     }
-                    None
-                })
-                .collect();
-            Ok(Signature::sign_single(
-                &mut t.clone(),
-                keys::aggregated_privkey(&signtx_keys)?,
-            ))
-        })
+                }
+                None
+            })
+            .collect();
+
+        let mut signtx_transcript = Transcript::new(b"ZkVM.signtx");
+        signtx_transcript.commit_bytes(b"txid", &utx.txid.0);
+        let sig = Signature::sign_multi(
```

Comment:
Will this `sign_multi` call also work if we have only one privkey, or do we need to call `sign_single` in that case?

Reply:
sign_multi will still work with one privkey. It uses the "multimessage API" (A.3 / defends against russel's attack), which is different from the "single message API" (normal musig), so calling sign_multi and sign_single are not interchangeable. (the verification for the two protocols is different.

Reply:
Multi-message signature with 1 key-message pair is inherently different (domain-separated) from the single message signature. In case of `signtx` instruction, we need to use multi-message API even if there's only one input, to keep things consistent.

Reply:
got it, that makes sense! thanks :)